examples/xds-management-server/pkg/snapshot/snapshot.go

```diff
@@ -0,0 +1,119 @@
+package snapshot
+
+import (
+	"context"
+	"time"
+
+	"quilkin.dev/xds-management-server/pkg/cluster"
+	"quilkin.dev/xds-management-server/pkg/filterchain"
+
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	log "github.com/sirupsen/logrus"
+
+	"quilkin.dev/xds-management-server/pkg/resources"
+)
+
+// RunSnapshotUpdater starts a goroutine that listens for resource updates,
+// uses the updates to generate an xds config snapshot and updates the snapshot from the provided channel and generates xds config snapshots.
+// cache with the latest snapshot for each connected node.
+// It returns the snapshot cache which can be passed on to the xds server impl.
+func RunSnapshotUpdater(
+	ctx context.Context,
+	logger *log.Logger,
+	clusterCh <-chan []cluster.Cluster,
+	filterChainCh <-chan filterchain.ProxyFilterChain,
+	updateInterval time.Duration,
+) cache.SnapshotCache {
+	snapshotCache := cache.NewSnapshotCache(false, cache.IDHash{}, logger)
+	go runSnapshotUpdater(
+		ctx,
+		logger,
+		clusterCh,
+		filterChainCh,
+		snapshotCache,
+		updateInterval)
+	return snapshotCache
+}
+
+// runSnapshotUpdater runs a loop that periodically checks if there are any
+// cluster/filter-chain updates and if so creates a snapshot for affected proxies
+// in the snapshot cache.
+func runSnapshotUpdater(
+	ctx context.Context,
+	logger *log.Logger,
+	clusterCh <-chan []cluster.Cluster,
+	filterChainCh <-chan filterchain.ProxyFilterChain,
+	snapshotCache cache.SnapshotCache,
+	updateInterval time.Duration,
+) {
+	logger = logger.WithFields(log.Fields{
+		"component": "SnapshotUpdater",
+	}).Logger
+
+	updateTicker := time.NewTicker(updateInterval)
```

Comment:
Non blocking suggestion: Could use something like github.com/jonboulle/clockwork to mock out time, so you can control the flow of time.